Data Structures/Trees/BST_insert-traverse-search-delete.cpp

```diff
@@ -0,0 +1,245 @@
+#include<iostream>
+using namespace std;
+
+//Single node of a tree
+struct node {
+    int data;
+    struct node* left;
+    struct node* right; 
+};
+
+//Class functions declaration
+class BST {
+    public:
+        struct node* insertion(int ,struct node*);
+        
+        void inorder(struct node*);
+        void preorder(struct node*);
+        void postorder(struct node*);
+
+        struct node* search(int, struct node*);
+
+        struct node* minValue(struct node*);
+        struct node* deletion(int, struct node*);
+};
+
+//-------------------DRIVER CODE--------------------
+
+int main() {
+
+    struct node *root = NULL;    //Initializing root as NULL
+    BST bst;
+
+    // The tree is like:
+    //     5
+    //     /\
+    //    4  9
+    //   /   /\
+    //  1   7  11
+    //   \    
+    //    3
+
+    root =  bst.insertion(5, root);    //First insertion becomes root
+    bst.insertion(4, root);
+    bst.insertion(9, root);
+    bst.insertion(1, root);
+    bst.insertion(7, root);
+    bst.insertion(11, root);
+    bst.insertion(3, root);
+
+    //This insertion is for duplicate node
+    bst.insertion(4,root);
+
+    //Tree traversal
+    cout<<"Displaying the tree:"<<endl;
+    cout<<"Inorder: "; bst.inorder(root);
+    cout<<endl;
+    cout<<"Preorder: "; bst.preorder(root);
+    cout<<endl;
+    cout<<"Postorder: "; bst.postorder(root);
+    cout<<endl;
+
+    //Searching for key
+    int key;
+    cout<<"Enter key to search: "; cin>>key;
+    bst.search(key, root);
+
+    //Deleting for key
+    cout<<"Enter key to delete: "; cin>>key;
+    root = bst.deletion(key, root);
+    
+    cout<<"Inorder: "; bst.inorder(root);    //Displaying tree in inorder after deletion
+    cout<<endl;
+
+    return 0;
+}
+
+//Function Definitions:
+
+//-------------------INSERTION--------------------
+
+//O(n) worst case, when a BST is in form of a linked list.
+//Average complexity is O(log n), as it keeps splitting the tree in left/right sub-trees
```

Comment:
Recurrence instead of "splitting the tree.."

Reply:
Added the requested changes. Please review.